fvm/state/address_generator.go

```diff
@@ -12,42 +12,66 @@ const keyAddressState = "account_address_state"
 // It uses the underlying generator it gets from the chain.
 // The only change is that when next address is called the state is updated as well.
 type StateBoundAddressGenerator struct {
-	generator flow.AddressGenerator
-	state     *State
+	stateHolder *StateHolder
+	chain       flow.Chain
 }
 
-func NewStateBoundAddressGenerator(state *State, chain flow.Chain) (*StateBoundAddressGenerator, error) {
-	stateBytes, err := state.Get("", "", keyAddressState)
+func NewStateBoundAddressGenerator(stateHolder *StateHolder, chain flow.Chain) (*StateBoundAddressGenerator, error) {
+	return &StateBoundAddressGenerator{
+		stateHolder: stateHolder,
+		chain:       chain,
+	}, nil
+}
+
+// TODO return error instead of a panic
+// this requires changes outside of fvm since the type is defined on flow model
+func (g *StateBoundAddressGenerator) Bytes() []byte {
```

Comment:
I think this method and subsequent ones are not expected to change the underlying `g` receiver. Hence, they may be better to be value receivers instead of pointer receivers.

Reply:
I honestly didn't touch this code that much, I think we need to change this heavily in an upcoming PR and we might use the error channel.

Reply:
we keep all the interfaces to be pointer in case they need to capture metrics and rate limit number of calls.